File: src/com/Torvald/ColourUtil/Col216.java
package com.Torvald.ColourUtil;

import org.newdawn.slick.Color;

/**
 * Created by minjaesong on 16-02-11.
 */
public class Col216 implements LimitedColours {

    private byte data;
    private static transient final int[] LOOKUP = {0x00, 0x33, 0x66, 0x99, 0xCC, 0xFF};

    public static final int MUL = 6;
    public static final int MUL_2 = MUL * MUL;
    public static final int MAX_STEP = MUL - 1;
    public static final int COLOUR_DOMAIN_SIZE = MUL_2 * MUL;

    /**
     *
     * @param data
     */
    public Col216(byte data) {
        create(data);
    }

    /**
     *
     * @param r 0-5
     * @param g 0-5
     * @param b 0-5
     */
    public Col216(int r, int g, int b) {
        create(r, g, b);
    }

    @Override
    public Color createSlickColor(int raw) {
        assertRaw(raw);
        int r = raw / MUL_2;
        int g = (raw % MUL_2) / MUL;
        int b = raw % MUL;

        return createSlickColor(r, g, b);
    }

    @Override
    public Color createSlickColor(int r, int g, int b) {
        assertRGB(r, g, b);
        return new Color(LOOKUP[r], LOOKUP[g], LOOKUP[b]);
    }

    @Override
    public void create(int raw) {
        assertRaw(raw);
        data = (byte) raw;
    }

    @Override
    public void create(int r, int g, int b) {
        assertRGB(r, g, b);
        data = (byte) (MUL_2 * r + MUL * g + b);
    }

    public byte getRaw() { return data; }

    private void assertRaw(int i) {
        if (i >= COLOUR_DOMAIN_SIZE || i < 0) {
            System.out.println("i: " + String.valueOf(i));
            throw new IllegalArgumentException();
        }
    }

    private void assertRGB(int r, int g, int b) {
        if (r > MAX_STEP || g > MAX_STEP || b > MAX_STEP || r < 0 || g < 0 || b < 0) {
            System.out.println("r: " + String.valueOf(r));
            System.out.println("g: " + String.valueOf(g));
            System.out.println("b: " + String.valueOf(b));
            throw new IllegalArgumentException();
        }
    }
}

File: src/com/Torvald/ColourUtil/Col216.kt
package com.torvald.colourutil

import org.newdawn.slick.Color

/**
 * 6-Step RGB with builtin utils.
 * Created by minjaesong on 16-02-11.
 */
class Col216 : LimitedColours {

    var raw: Byte = 0
        private set

    /**

     * @param data
     */
    constructor(data: Byte) {
        create(data.toInt())
    }

    /**

     * @param r 0-5
     * *
     * @param g 0-5
     * *
     * @param b 0-5
     */
    constructor(r: Int, g: Int, b: Int) {
        create(r, g, b)
    }

    override fun createSlickColor(raw: Int): Color {
        assertRaw(raw)
        val r = raw / MUL_2
        val g = raw % MUL_2 / MUL
        val b = raw % MUL

        return createSlickColor(r, g, b)
    }

    override fun createSlickColor(r: Int, g: Int, b: Int): Color {
        assertRGB(r, g, b)
        return Color(LOOKUP[r], LOOKUP[g], LOOKUP[b])
    }

    override fun create(raw: Int) {
        assertRaw(raw)
        this.raw = raw.toByte()
    }

    override fun create(r: Int, g: Int, b: Int) {
        assertRGB(r, g, b)
        raw = (MUL_2 * r + MUL * g + b).toByte()
    }

    private fun assertRaw(i: Int) {
        if (i >= COLOUR_DOMAIN_SIZE || i < 0) {
            println("i: " + i.toString())
            throw IllegalArgumentException()
        }
    }

    private fun assertRGB(r: Int, g: Int, b: Int) {
        if (r !in 0..MAX_STEP || g !in 0..MAX_STEP || b !in 0..MAX_STEP) {
            println("r: " + r.toString())
            println("g: " + g.toString())
            println("b: " + b.toString())
            throw IllegalArgumentException()
        }
    }

    companion object {
        @Transient private val LOOKUP = intArrayOf(0x00, 0x33, 0x66, 0x99, 0xCC, 0xFF)

        const val MUL = 6
        const val MUL_2 = MUL * MUL
        const val MAX_STEP = MUL - 1
        const val COLOUR_DOMAIN_SIZE = MUL_2 * MUL
    }
}
